File: PaymentLibrary-Android/PaymentLibrary/src/main/java/com/dynatracese/paymentlibrary/PaymentClient.kt
//=============================================================================
// src/main/java/com/dynatracese/paymentlibrary/PaymentClient.kt
//
// Esta classe é a interface pública da biblioteca.
// Ela expõe as funcionalidades de receber e cancelar pagamentos.
//=============================================================================
package com.dynatracese.paymentlibrary

import android.util.Log
import android.os.Build
//import com.dynatracese.paymentlibrary.BuildConfig
import com.dynatracese.paymentlibrary.api.PaymentService
import com.dynatracese.paymentlibrary.models.PaymentCancellationRequest
import com.dynatracese.paymentlibrary.models.PaymentRequest
import com.dynatracese.paymentlibrary.models.PaymentResponse
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.coroutines.delay
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.lang.Exception
import java.util.UUID
import android.content.Context
import android.widget.Toast
import kotlin.random.Random
import com.dynatracese.paymentlibrary.PaymentCrashHandler
import java.net.Inet4Address
import java.net.NetworkInterface
import java.util.Collections
import java.io.PrintWriter
import java.io.StringWriter

// --- FIX: REMOVED UNNECESSARY IMPORTS ---
// Classes in the same package are automatically visible.
// ---------------------------------------


// Interface para callbacks de pagamento
interface PaymentCallback {
    fun onPaymentSuccess(transactionId: String)
    fun onPaymentFailure(error: String)
}

// Interface para callbacks de cancelamento
interface CancellationCallback {
    fun onCancellationSuccess()
    fun onCancellationFailure(error: String)
}

// A classe agora recebe um objeto Config e o Context
class PaymentClient private constructor(
    private val config: Config, // Use a Config object
    private val context: Context
) {

    /**
     * Unified configuration for the PaymentClient
     * @param paymentBaseUrl The base URL for the payment processing backend
     * 
     * Note: Library version (appVersion) is automatically collected from BuildConfig
     * and cannot be overridden to ensure proper crash deobfuscation in Dynatrace.
     */
    data class Config(
        val paymentBaseUrl: String
    )

    companion object {
        @Volatile
        private var INSTANCE: PaymentClient? = null

        // Updated getInstance to accept the new Config object
        fun getInstance(config: Config, context: Context): PaymentClient {
            // Use the application context to prevent memory leaks
            val appContext = context.applicationContext
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: PaymentClient(config, appContext).also { INSTANCE = it }
            }
        }
    }

    private val paymentService: PaymentService?

    init {
        // Register crash handler
        PaymentCrashHandler.register(context.applicationContext)

        // Automatically collect library version from BuildConfig (not from host app)
        val libraryVersion = BuildConfig.VERSION_NAME

        // Initialize the BusinessEventsClient using the new Secrets-aware method
        BusinessEventsClient.configureFromSecrets(
            context = context,
            eventProvider = "Android App", // Internal default, not customizable
            appVersion = libraryVersion // Always use library version, not app version
        )

        // Se a URL for "TEST_ONLY", não inicializa o Retrofit
        // Use config.paymentBaseUrl instead of baseUrl
        if (config.paymentBaseUrl != "TEST_ONLY") {
            val retrofit = Retrofit.Builder()
                .baseUrl(config.paymentBaseUrl) // Use config
                .addConverterFactory(GsonConverterFactory.create())
                .build()

            paymentService = retrofit.create(PaymentService::class.java)
        } else {
            paymentService = null
        }

        DynatraceLogger.info("PaymentLibrary initialized successfully", "PaymentClient")
    }

    /**
     * Recebe um pagamento e retorna o ID da transação.
     * @param amount Valor do pagamento.
     * @param creditCardNumber Número do cartão de crédito.
     * @param vendorName Nome do fornecedor.
     * @param vendorId ID do fornecedor.
     */
    suspend fun receivePayment(
        amount: Double,
        creditCardNumber: String,
        vendorName: String,
        vendorId: String,
        callback: PaymentCallback,
        crashStatus: Boolean
    ) {
        // Use BusinessEventsClient to track the payment process
        BusinessEventsClient.withAction(
            name = "receivePayment",
            attributes = mapOf(
                "payment.amount" to amount,
                "payment.creditCardNumber" to creditCardNumber,
                "payment.vendorName" to vendorName,
                "payment.vendorId" to vendorId
            )
        ) {
            DynatraceLogger.info("Starting payment processing", "PaymentClient")

            if (crashStatus) {
                DynatraceLogger.error("Simulating crash on purpose", "PaymentClient")
                throw Exception("Simulated Payment Library Crash")
            }

            // Execute both executePayment and dummyDoSomething in parallel
            coroutineScope {
                val paymentDeferred = async {
                    executePayment(amount, creditCardNumber, vendorName, vendorId, callback)
                }

                val dummyDeferred = async {
                    dummyDoSomething()
                }

                // Wait for both operations to complete
                paymentDeferred.await()
                dummyDeferred.await()
            }

            // Add random delay like iOS version
            val randomDelayMs = (700..3000).random().toLong()
            delay(randomDelayMs)
        }
    }

    private fun getLocalIpAddress(): String? {
        try {
            val interfaces = Collections.list(NetworkInterface.getNetworkInterfaces())
            for (intf in interfaces) {
                val addrs = Collections.list(intf.inetAddresses)
                for (addr in addrs) {
                    if (!addr.isLoopbackAddress && addr is Inet4Address) {
                        return addr.hostAddress
                    }
                }
            }
        } catch (ex: Exception) {
            // Handle exception silently
        }
        return null
    }

    private suspend fun executePayment(
        amount: Double,
        creditCardNumber: String,
        vendorName: String,
        vendorId: String,
        callback: PaymentCallback
    ) {
        BusinessEventsClient.withAction(
            name = "executePayment",
            attributes = mapOf(
                "payment.amount" to amount,
                "payment.creditCardNumber" to creditCardNumber,
                "payment.vendorName" to vendorName,
                "payment.vendorId" to vendorId
            )
        ) {
            val randomDelayMs = (700..3000).random().toLong()
            delay(randomDelayMs)

            // Use config.paymentBaseUrl
            if (config.paymentBaseUrl == "TEST_ONLY") {
                if (amount > 0) {
                    val transactionId = UUID.randomUUID().toString()
                    callback.onPaymentSuccess(transactionId)
                } else {
                    callback.onPaymentFailure("Amount must be positive")
                    Log.w("executePayment", "⚠️ Payment failed in TEST_ONLY mode: Amount not positive")
                }
                return@withAction
            }

            withContext(Dispatchers.IO) {
                try {
                    val paymentRequest = PaymentRequest(amount, creditCardNumber, vendorName, vendorId)
                    val response = paymentService?.receivePayment(paymentRequest)

                    if (response != null && response.isSuccessful) {
                        val transactionId = response.body()?.transactionId
                        if (transactionId != null) {
                            callback.onPaymentSuccess(transactionId)
                        } else {
                            callback.onPaymentFailure("Transaction ID not found in response")
                            Log.e("executePayment", "❌ Failed to get transaction ID from response")
                        }
                    } else {
                        val statusCode = response?.code() ?: 0
                        callback.onPaymentFailure("HTTP $statusCode")
                        Log.e("executePayment", "❌ Payment API call failed with HTTP status code: $statusCode")
                    }
                } catch (e: Exception) {
                    callback.onPaymentFailure(e.message ?: "Unknown error")
                    Log.e("executePayment", "❌ Payment network request failed: ${e.message}")
                }
            }
        }
    }

    private suspend fun dummyDoSomething() {
        BusinessEventsClient.withAction(name = "dummyDoSomething") { // Renamed action
            val randomDelayMs = (700..3000).random().toLong()
            delay(randomDelayMs)
            dummyDoSomethingElse()
        }
    }

    private suspend fun dummyDoSomethingElse() {
        BusinessEventsClient.withAction(name = "dummyDoSomethingElse") { // Renamed action
            val randomDelayMs = (700..3000).random().toLong()
            delay(randomDelayMs)
            dummyDoSomethingMore()
        }
    }

    private suspend fun dummyDoSomethingMore() {
        BusinessEventsClient.withAction(name = "dummyDoSomethingMore") {
            val randomDelayMs = (700..3000).random().toLong()
            delay(randomDelayMs)
        }
    }

    /**
     * Cancela um pagamento.
     * @param transactionId ID da transação a ser cancelada.
     * @param callback Callback para tratar o sucesso ou a falha.
     */
    suspend fun cancelPayment(
        transactionId: String,
        callback: CancellationCallback
    ) {
        // Use config.paymentBaseUrl
        withContext(Dispatchers.IO) {
            if (config.paymentBaseUrl == "TEST_ONLY") {
                // Simulação de cancelamento para o modo de teste
                if (transactionId.isNotEmpty()) {
                    callback.onCancellationSuccess()
                } else {
                    callback.onCancellationFailure("Simulated error: Transaction ID is empty.")
                }
            } else {
                // Lógica de chamada real para o backend
                try {
                    val cancellationRequest = PaymentCancellationRequest(transactionId)
                    val response = paymentService?.cancelPayment(cancellationRequest)

                    if (response != null && response.isSuccessful) {
                        callback.onCancellationSuccess()
                    } else {
                        callback.onCancellationFailure("Cancellation failed with code: ${response?.code()}")
                    }
                } catch (e: Exception) {
                    Log.e("PaymentClient", "❌ Error canceling payment", e)
                    callback.onCancellationFailure(e.message ?: "Unknown error")
                }
            }
        }
    }
}

File: PaymentLibrary-iOS/PaymentLibrary/PaymentLibrary.docc/PaymentClient.swift
// PaymentClient.swift — dentro do módulo PaymentLibrary

import Foundation
import os.log
#if canImport(UIKit)
import UIKit
#endif

// MARK: - Callbacks (públicos)
public protocol PaymentCallback: AnyObject {
    func onPaymentSuccess(transactionId: String)
    func onPaymentFailure(error: String)
}

public protocol CancellationCallback: AnyObject {
    func onCancellationSuccess()
    func onCancellationFailure(error: String)
}

// MARK: - Modelos internos (não expostos)
struct PaymentRequest: Encodable {
    let amount: Double
    let creditCardNumber: String
    let vendorName: String
    let vendorId: String
}

struct PaymentResponse: Decodable {
    let transactionId: String
}

struct PaymentCancellationRequest: Encodable {
    let transactionId: String
}

// MARK: - Cliente
public final class PaymentClient {

    // Singleton
    public static var shared: PaymentClient {
        guard let instance = instance else {
            fatalError("PaymentClient not initialized. Call getInstance(baseUrl:) first.")
        }
        return instance
    }
    private static var instance: PaymentClient?

    private let baseUrl: String

    /// Inicializa (uma vez) o singleton
    @discardableResult
    public static func getInstance(baseUrl: String) -> PaymentClient {
        if let i = instance { return i }
        let i = PaymentClient(baseUrl: baseUrl)
        instance = i
        return i
    }

    private init(baseUrl: String) {
        self.baseUrl = baseUrl
        #if canImport(UIKit)
        os_log("PaymentClient init iOS %@", type: .info, UIDevice.current.systemVersion)
        #else
        os_log("PaymentClient init", type: .info)
        #endif
    }

    // MARK: - API pública
    public func receivePayment(
        amount: Double,
        creditCardNumber: String,
        vendorName: String,
        vendorId: String,
        callback: PaymentCallback,
        crashStatus: Bool
    ) {
        os_log("Starting payment process.", type: .info)

        if crashStatus {
            os_log("Simulating crash.", type: .error)
            fatalError("Simulated Payment Library Crash")
        }

        Task {
            await executePayment(
                amount: amount,
                creditCardNumber: creditCardNumber,
                vendorName: vendorName,
                vendorId: vendorId,
                callback: callback
            )
        }
    }

    public func cancelPayment(
        transactionId: String,
        callback: CancellationCallback
    ) {
        Task { await executeCancellation(transactionId: transactionId, callback: callback) }
    }

    // MARK: - Internals
    private func executePayment(
        amount: Double,
        creditCardNumber: String,
        vendorName: String,
        vendorId: String,
        callback: PaymentCallback
    ) async {
        if baseUrl == "TEST_ONLY" {
            if amount > 0 {
                DispatchQueue.main.async {
                    callback.onPaymentSuccess(transactionId: UUID().uuidString)
                }
            } else {
                DispatchQueue.main.async {
                    callback.onPaymentFailure(error: "Amount must be positive.")
                }
            }
            return
        }

        guard let url = URL(string: baseUrl + "/receive_payment") else {
            DispatchQueue.main.async {
                callback.onPaymentFailure(error: "Invalid base URL")
            }
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let body = PaymentRequest(amount: amount, creditCardNumber: creditCardNumber, vendorName: vendorName, vendorId: vendorId)

        do {
            request.httpBody = try JSONEncoder().encode(body)
            let (data, response) = try await URLSession.shared.data(for: request)
            guard (response as? HTTPURLResponse)?.statusCode == 200 else {
                DispatchQueue.main.async {
                    callback.onPaymentFailure(error: "HTTP \((response as? HTTPURLResponse)?.statusCode ?? 0)")
                }
                return
            }
            let decoded = try JSONDecoder().decode(PaymentResponse.self, from: data)
            DispatchQueue.main.async { callback.onPaymentSuccess(transactionId: decoded.transactionId) }
        } catch {
            DispatchQueue.main.async { callback.onPaymentFailure(error: error.localizedDescription) }
        }
    }

    private func executeCancellation(
        transactionId: String,
        callback: CancellationCallback
    ) async {
        if baseUrl == "TEST_ONLY" {
            DispatchQueue.main.async {
                transactionId.isEmpty ? callback.onCancellationFailure(error: "Empty transactionId")
                                      : callback.onCancellationSuccess()
            }
            return
        }

        guard let url = URL(string: baseUrl + "/cancel_payment") else {
            DispatchQueue.main.async { callback.onCancellationFailure(error: "Invalid base URL") }
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let body = PaymentCancellationRequest(transactionId: transactionId)

        do {
            request.httpBody = try JSONEncoder().encode(body)
            let (_, response) = try await URLSession.shared.data(for: request)
            guard (response as? HTTPURLResponse)?.statusCode == 200 else {
                DispatchQueue.main.async {
                    callback.onCancellationFailure(error: "HTTP \((response as? HTTPURLResponse)?.statusCode ?? 0)")
                }
                return
            }
            DispatchQueue.main.async { callback.onCancellationSuccess() }
        } catch {
            DispatchQueue.main.async { callback.onCancellationFailure(error: error.localizedDescription) }
        }
    }
}
